content/blog/2020-08-05-cml-self-hosted-runners-on-demand-with-gpus.md

```diff
@@ -0,0 +1,183 @@
+    With the containerised runner you are restricting the access to your real
+    machine.
+
+4.  Gain reproducibility. One of the biggest technical debts in the ML space is
+    reproducibility. A few weeks post-experiment, we often discover that trying
+    to put your model back in shape is a pain. Looking at our repo, it’s not
+    obvious what data or training infrastructure or dependencies went into a
+    given result. When you move your ML experiments into a CI/CD system you are
+    making a contract of the dependencies and hardware used for your experiment.
+    Having that contract isolated by the containerised runner, your experiment
+    is perfectly reproducible by anyone in the future.
+
+## Hands on GPU Self-hosted runners 101
+
+### 1) Install nvidia drivers and nvidia-docker in your machine (ubuntu 18.04)
+
+```dvc
+$ curl -s -L https://nvidia.GitHub.io/nvidia-docker/gpgkey | sudo apt-key add - && \
+  curl -s -L https://nvidia.GitHub.io/nvidia-docker/ubuntu18.04/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list && \
+  sudo apt update && sudo apt install -y ubuntu-drivers-common  && \
+  sudo ubuntu-drivers autoinstall  && \
+  sudo apt install -y nvidia-container-toolkit && \
+  sudo systemctl restart docker
+```
+
+You can test that your gpus are up and running with the following command:
+
+```dvc
+$ docker run --gpus all dvcorg/cml-py3 nvidia-smi
+```
+
+We should see something like this:
+![](/uploads/images/2020-08-05/nvidia-smi-output.png)
+
+### 2) Start your self-hosted runner
+
+With CML docker images launching your own self-hosted runner is very easy. They
+comes in two flavours, depending if you need python 2 or 3:
+
+- dvcorg/cml:latest
+- dvcorg/cml-py3:latest
+
+These images have CML and DVC preinstalled (among other perks), plus CUDA
+drivers and Python 2.7. The cml-py3 image comes with Python 3.X as well. That's
+all. You can clone these images and add your own dependencies to better mimic
+your own production environment.
+
+```dvc
+$ docker run --name myrunner -d --gpus all \
+    -e RUNNER_IDLE_TIMEOUT=1800 \
+    -e RUNNER_LABELS=cml,gpu \
+    -e RUNNER_REPO=$my_repo_url \
+    -e repo_token=$my_repo_token \
+    dvcorg/cml-py3
+```
+
+where:
+
+`RUNNER_IDLE_TIMEOUT` is the time in seconds that the runner is going to be idle
+at most waiting for jobs to come, if no one comes the runner shuts down and
+unregisters from your repo.
+
+`RUNNER_LABELS` a comma delimited list of labels that we are setting in our
+workflow that the jobs will wait for.
+
+`RUNNER_REPO` is the url of your GitLab or GitHub repo. repo_token is the
+personal token generated for your GitHub or GitLab repo. Note that for GitHub
+you must check `workflow` along with `repo`.
+
+If everything went fine we should see a runner registered in our repo.
+
+![](/uploads/images/2020-08-05/registered-cml-runner-GitHub.png)
```

Comment:
that's what I see:

<img width="712" alt="Screen Shot 2020-08-07 at 10 25 41 AM" src="https://user-images.githubusercontent.com/3659196/89671852-8e2a8000-d898-11ea-9894-e826aeca49ee.png">

Reply:
@shcheklein this is what I see

![image](https://user-images.githubusercontent.com/414967/89672150-898bb900-d8e4-11ea-929c-7177f763d2f6.png)

Reply:
ah! 
got it!

Reply:
its case sensitive @shcheklein, my bad not having tested this on linux vs macos